dash/development/_jl_components_generation.py

```diff
@@ -0,0 +1,523 @@
+            end
+
+            push!(result.props, prop => Front.to_dash(value))
+        end
+
+    return result
+end
+
+function {funcname}(children::Any; kwargs...)
+    result = {funcname}(;kwargs...)
+    push!(result.props, :children => Front.to_dash(children))
+    return result
+end
+
+{funcname}(children_maker::Function; kwargs...) = {funcname}(children_maker(); kwargs...)
+'''  # noqa:E501
+
+jl_package_file_string = '''
+module {package_name}
+using Dash
+
+const resources_path = realpath(joinpath( @__DIR__, "..", "deps"))
+const version = "{version}"
+
+{component_includes}
+
+function __init__()
+    Dash.register_package(
+        Dash.ResourcePkg(
+            "{project_shortname}",
+            resources_path,
+            version = version,
+            [
+                {resources_dist}
+            ]
+        )
+    )
+end
+end
+'''
+
+jl_projecttoml_string = '''
+name = "{package_name}"
+uuid = "{package_uuid}"
+{authors}
+version = "{version}"
+
+[deps]
+Dash = "{dash_uuid}"
+
+[compact]
+julia = "1.1"
+Dash = ">=0.1"
+'''
+
+jl_component_include_string = 'include("{name}.jl")'
+
+jl_resource_tuple_string = '''Dash.Resource(
+    relative_package_path = {relative_package_path},
+    external_url = {external_url},
+    dynamic = {dynamic},
+    async = {async_string},
+    type = :{type}
+)'''
+
+
+def jl_package_name(namestring):
+    s = namestring.split("_")
+    return "".join(w.capitalize() for w in s)
+
+
+def stringify_wildcards(wclist, no_symbol=False):
+    if no_symbol:
+        wcstring = "|".join(
+            '{}-'.format(item) for item in wclist
+        )
+    else:
+        wcstring = ", ".join(
+            'Symbol("{}-")'.format(item) for item in wclist
+        )
+    return wcstring
+
+
+def get_wildcards_jl(props):
+    prop_keys = list(props.keys())
+    wildcards = []
+    for key in prop_keys:
+        if key.endswith("-*"):
+            wildcards.append(key.replace("-*", ""))
+    return wildcards
+
+
+def get_jl_prop_types(type_object):
+    """Mapping from the PropTypes js type object to the Julia type."""
+
+    def shape_or_exact():
+        return "lists containing elements {}.\n{}".format(
+            ", ".join("'{}'".format(t) for t in list(type_object["value"].keys())),
+            "Those elements have the following types:\n{}".format(
+                "\n".join(
+                    create_prop_docstring_jl(
+                        prop_name=prop_name,
+                        type_object=prop,
+                        required=prop["required"],
+                        description=prop.get("description", ""),
+                        indent_num=1,
+                    )
+                    for prop_name, prop in list(type_object["value"].items())
+                )
+            ),
+        )
+
+    return dict(
+        array=lambda: "Array",
+        bool=lambda: "Bool",
+        number=lambda: "Float64",
```

Comment:
I'm not sure if there are functional consequences of this, but at least from a documentation perspective might it be better to include more number types? `Float64 | Int64` looks like it covers the output of `JSON2.read`, ie values to be received by callback functions. But values you can send back as callback returns are broader, likely `Real`.

Reply:
This is only used in documentation. Yes, it makes sense to replace types with abstract ones. 
Abstract types for numbers looks like this:
```
abstract type Number end
abstract type Real     <: Number end
abstract type AbstractFloat <: Real end
abstract type Integer  <: Real end
abstract type Signed   <: Integer end
abstract type Unsigned <: Integer end
```

Reply:
This is only used in documentation. Yes, it makes sense to replace types with abstract ones.
Abstract types for numbers looks like this:
```
abstract type Number end
abstract type Real     <: Number end
abstract type AbstractFloat <: Real end
abstract type Integer  <: Real end
abstract type Signed   <: Integer end
abstract type Unsigned <: Integer end
```

Reply:
OK good, let's go with `Real`.

Reply:
There's also the `any=lambda: "Bool | Float64 | String | Dict | Array"` below.  I remember finding these annotations very helpful when looking at the online Python docs.  And we can express this directly in Julia with `Union{...}`s.

In fact, it looks like we can use the `union` lambda below to print out real Julian `Union{}`s using your existing architecture.  I'm afraid it'd take further restructuring to make [an options block like this](https://dash.plotly.com/dash-core-components/checklist) format like the below for Julia, but this would be closer to how we'd naturally document these things:

> * `label::Union{String, Number}` (required): The checkbox's label
> * `value::Union{String, Number}` (required): The value of the checkbox. This value corresponds to the items specified in the value property.
> * `disabled::Bool` (optional): If true, this checkbox is disabled and can't be clicked on.

All that said, this sort of strict notation doesn't work as well for arrays and dicts, though, since we support far more than just `Array` and `Dict`.  The builtin `Array` is just one of many possible `AbstractArray`s, and JSON2 also supports serializing `Tuple`s and `Set`s the same way.  Similarly, the builtin `Dict` is just one of many dictionary-like things; JSON2 also supports NamedTuples and even your own structs.  That said, I don't want to lean too heavily on what the particular JSON library does.  This gets into the nitty gritty, but JSON2 should probably serialize all `AbstractDict`s as JSON dicts; it [doesn't right now](https://github.com/quinnj/JSON2.jl/issues/35). It may be good to move to JSON3 in the future which has a [far more robust manner of flagging these types](https://quinnj.github.io/JSON3.jl/stable/) and is even speedier, so tying yourself to a particular implementation isn't the best.

One way we distinguish between the concrete implementations (like `Array` and `Dict`) and their general behaviors in our own documentation is just to use lowercase without code formatting, simply: array and dictionary.  And if I were using these sorts of more relaxed terms, I wouldn't use the `Union{}` notation at all.

So all that to say, maybe don't uppercase any of these guys?  Any maybe keep the (x | y | z) union printing?

Reply:
The point is that these types are actually mapping what JS expects and the question of linking them to Julia types is fairly conditional. I don't know how to write this in the documentation to make it clear. There is no verification of these types at the Julia level. To json dict also mapped NamedTuple, for example and Tuple mapped to json array.

Reply:
>So all that to say, maybe don't uppercase any of these guys? Any maybe keep the (x | y | z) union printing?

I support it. With an indication in the documentation that these are types at the js level, not at the Julia level

Reply:
> these are types at the js level, not at the Julia level

Yes, but the Julia app developer needs to know two things:
- What type(s) can they expect to receive in callbacks?
- What type(s) can they provide as prop values, either as callback return values or in layout components?

The more precision we can provide around these two issues the better. The first is easier, as the set of possible inputs from a JSON string is fairly limited. And then I assume that JSON2 is at least symmetric in the sense that `JSON2.write(JSON2.read(s))==s` - ie if you provide a return value consistent with the argument value we would provide, it will always work.

Given that, I'd prefer that we base our component documentation (ie this generated code) on this more restricted set of types you get from de-serializing JSON, and then somewhere in our general `dash.jl` docs we can describe the other types that you can provide as prop values - and we can frame this explicitly as "Anything that JSON2 (or JSON3 if we switch to that) will serialize the same way as type X."

In Dash for Python we've found that folks get confused when these types don't match - for example the DCC date pickers can be initialized using Python `datetime` objects, because they have a standard JSON serialization, but when you create a callback it will receive the date as a string. So we document this as a string type and explicitly stringify it in our [main example](https://dash.plotly.com/dash-core-components/datepickersingle):
```py
date=str(dt(2017, 8, 25, 23, 59, 59))
```
(though I notice in later examples you'll see a `datetime` passed in directly - we should probably switch them all to strings, and/or make an explicit note about serializing `datetime` objects!)

Reply:
You should also keep in mind that each argument is converted to JSON using `Front.to_dash`. By default, this is an identical conversion. But it can be overloaded with third-party packages. For example you can use it to pass a PlotlyBase object instead of dcc_graph

Reply:
@alexcjohnson changed to `Real` in fd24e508

Reply:
As @mbauman mentioned we should also update this in `any`

Reply:
fixed in 7eb50221